# Profitable Mobile Apps for App Store and Google Play Markets

For this project, we will analyze free mobile apps on Google Play and the App Store to find out which ones are more attractive for users. Since our apps are free to download and install, this means and our main source of revenue consists of in-app ads. The more users that see and engage with the ads, the most profitable the app is being become for us. 

Our goal for this project is to analyze data to understand what type of apps attract more users.

# Opening and Exploring the Data

As of September 2018, there were approximately 2 million iOS apps available on the App Store, and 2.1 million Android apps on Google Play.

Collecting data for over four million apps requires a significant amount of time and money, so we'll try to analyze a sample of data instead. To avoid spending resources with collecting new data ourselves, we should first try to see whether we can find any relevant existing data at no cost. Luckily, these are two data sets that seem suitable for our purpose:

* [First data set](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) containing data about approximately 7,000 iOS apps from the App Store; the data was collected in July 2017.
* [Second data set](https://www.kaggle.com/lava18/google-play-store-apps) containing data about approximately 10,000 Android apps from Google Play; the data was collected in August 2018.

Let's start by opening the two data sets and then continue with exploring the data.

In [1]:
opened_file = open('AppleStore.csv', encoding="UTF-8")
from csv import reader
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

In [2]:
opened_file = open('googleplaystore.csv', encoding="UTF-8")
from csv import reader
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

To make it easier to explore the two data sets, we'll first write a function named `explore_data()` that we can use repeatedly to explore rows in a more readable way. We'll also add an option for our function to show the number of rows and columns for any data set.

In [3]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n')
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

## iOS apps from the App Store

In [4]:
print(ios_header)
print('\n')
explore_data(ios, 0, 3, rows_and_columns=True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


We have **7,197** iOS apps in this data set, and the columns that seem interesting are: 
- 'track_name', 
- 'currency', 
- 'price', 
- 'rating_count_tot', 
- 'rating_count_ver',
- 'prime_genre'

Not all column names are self-explanatory in this case, but details about each column can be found in the data set [documentation](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps).

## Android apps from Google Play

In [5]:
print(android_header)
print('\n')
explore_data(android, 0, 4, rows_and_columns=True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 10841
Number of columns: 13


We have **10,841** Android apps in this data set, and the columns that seem interesting are: 
- 'App', 
- 'Category', 
- 'Rating', 
- 'Type', 
- 'Price',
- 'Installs'
- 'Genres'

Not all column names are self-explanatory in this case, but details about each column can be found in the data set [documentation](https://www.kaggle.com/lava18/google-play-store-apps).

# Deleting Wrong Data

The Google Play data set has a dedicated [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion), and we can see that [one of the discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) outlines an error for row 10472 or 10473 depending on whether the android apps list has a header row or not. Let's explore these rows and check which one is with the error for our android apps list that includes a header row.

In [6]:
explore_data(android, 10472, 10475)

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


['Sat-Fi Voice', 'COMMUNICATION', '3.4', '37', '14M', '1,000+', 'Free', '0', 'Everyone', 'Communication', 'November 21, 2014', '2.2.1.5', '2.2 and up']




Above, we can see that in the row 10472 missing value for the 'Category' column comparing to other rows. As a result, data have been shifted, so the rating for `Life Made WI-Fi Touchscreen Photo Frame` app is `19` that's not correct because the maximum rating for a Google Play app is 5. As a consequence, we'll delete this row.

In [7]:
del android[10472]
explore_data(android, 10472, 10475)

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


['Sat-Fi Voice', 'COMMUNICATION', '3.4', '37', '14M', '1,000+', 'Free', '0', 'Everyone', 'Communication', 'November 21, 2014', '2.2.1.5', '2.2 and up']


['Wi-Fi Visualizer', 'TOOLS', '3.9', '132', '2.6M', '50,000+', 'Free', '0', 'Everyone', 'Tools', 'May 17, 2017', '0.0.9', '2.3 and up']




## Attension!

Don't run above code cell more than once, otherwise you'll delete more than one row. Above, we can see that original row 10472 was deleted and now under this number there is previous 10473 row with `osmino Wi-Fi: free WiFi` app.
Let's now check new length for android list without header row, recall that before deleting it was **10,841**.

In [8]:
print(len(android))

10840


# Deleting duplicate entries

## Explanation.

Exploring the Google Play data set long enough or looking at the mentioned above [discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion) section, we'll notice some apps have duplicate entries. 
For instance, `Instagram` has four entries, see below.

In [9]:
for app in android:
    name=app[0]
    if name=='Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


Next, let's check the overall qty of duplicate entries for android apps.

In [10]:
duplicate_apps=[]
unique_apps=[]
for app in android:
    name=app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

print('Overall qty of duplicate entries for android apps is: ', len(duplicate_apps))
print('\n')
print('Overall qty of unique entries for android apps is: ', len(unique_apps))
print('\n')
print('Examples of duplicate apps:   ',duplicate_apps[:10])

Overall qty of duplicate entries for android apps is:  1181


Overall qty of unique entries for android apps is:  9659


Examples of duplicate apps:    ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


We have **9,659** unique entries and **1,181** duplicate entries for android apps, but how to decide which ones should be removed and which left? 

For example, if we look at the `Instagram` duplicates, we can notice that the only different between them is for `Reviews` numbers.

![Image](https://i.imgur.com/5j2savp.png)

So, the higher the number of reviews, the more recent the data should be. Rather than removing duplicates randomly, we'll only keep the row with the highest number of reviews because the higher the number of reviews, the more reliable the ratings.

To do that, we will:

* Create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app
* Use the dictionary to create a new data set, which will have only one entry per app (and we only select the apps with the highest number of reviews)

## Cleaning.

In [11]:
reviews_max={}
for app in android:
    name=app[0]
    n_reviews=float(app[3])
    if name not in reviews_max:
        reviews_max[name]=n_reviews
    elif name in reviews_max and n_reviews>reviews_max[name]:
        reviews_max[name]=n_reviews
print('Dictionary length is: ', len(reviews_max))

Dictionary length is:  9659


We see that the length of the created dictionary is **9,659** apps that's the same as the number of unique Android apps. As a result, the created dictionary includes only unique Android apps with the highest reviews number.

Now, let's use the `reviews_max` dictionary to remove the duplicates. In the code cell below:

* We start by initializing two empty lists, `android_clean` and `already_added`.
* We loop through the android data set, and for every iteration:
    - We isolate the name of the app and the number of reviews.
    - We add the current row (`app`) to the `android_clean` list, and the app name (`name`) to the `already_cleaned` list if:
        - The number of reviews of the current app matches the number of reviews of that app as described in the `reviews_max` dictionary; and
        - The name of the app is not already in the `already_added` list. We need to add this supplementary condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the `Box` app has three entries, and the number of reviews is the same). If we just check for `reviews_max[name] == n_reviews`, we'll still end up with duplicate entries for some apps.

In [12]:
android_clean=[]
already_added=[]
for app in android:
    name=app[0]
    n_reviews=float(app[3])
    if reviews_max[name]==n_reviews and name not in already_added:
        android_clean.append(app)
        already_added.append(name)
print('Length of cleaned Android apps list is: ', len(android_clean))
print('Overall number of unique apps is: ', len(already_added))
print('\n')
print('Examples of unique apps in the new data set:')
explore_data(android_clean, 0, 4)

Length of cleaned Android apps list is:  9659
Overall number of unique apps is:  9659


Examples of unique apps in the new data set:
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']




As we see above, new cleaned data set has **9,659** rows, just as expected.

# Removing Non-English Apps

## Investigation & Identification

If you explore the data long enough, you'll find that both data sets - `android` and `ios` - have apps with names that suggest they are not directed toward an English-speaking audience. Below are samples for both data sets.

In [13]:
print (android_clean[4412][0])
print (android_clean[7940][0])
print('\n')
print (ios[814][1])
print (ios[6732][1])

中国語 AQリスニング
لعبة تقدر تربح DZ


Filterra – Photo Editor, Effects for Pictures
クリスマスミッケ/脱出ゲーム感覚の絵探しパズルゲーム


Each character we use in a string has a corresponding number associated with it. For instance, the corresponding number for character 'a' is 97, character 'A' is 65, and character '爱' is 29,233.

In [14]:
print(ord('a'))
print(ord('A'))
print(ord('爱'))

97
65
29233


The numbers corresponding to the characters we commonly use in an English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;), and other symbols (+, *, /) are all in the range 0 to 127, according to the [ASCII](https://en.wikipedia.org/wiki/ASCII) (American Standard Code for Information Interchange). 

Based on this number range, we can build a function that detects whether a character belongs to the set of common English characters or not. If the number is equal to or less than 127, then the character belongs to the set of common English characters.

In Python, strings are indexable and iterable, which means we can use indexing to select an individual character, and we can also iterate on the string using a for loop.

In [15]:
def is_english(name):
    for character in name:
        if ord(character) > 127:
            return False
    return True

In [16]:
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
False
False


As we see, the function couldn't correctly identify certain English app names like *'Docs To Go™ Free Office Suite'* and *'Instachat 😜'*. This is because emojis and characters like ™ fall outside the ASCII range and have corresponding numbers over 127. 

In [17]:
print(ord('™'))
print(ord('😜'))

8482
128540


Not to lose good names we'll modify a little our created function. So the rule will be next: if the app's name has more than three characters with corresponding numbers falling outside the ASCII range, then it will be removed. This means all English apps with up to three emoji or other special characters will still be labeled as English.

In [18]:
def is_english(name):
    non_english=[]
    for character in name:
        if ord(character) > 127:
            non_english.append(character)
    if len(non_english) > 3:
        return False
    else:
        return True

In [19]:
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


## Cleaning 2.

The function is still not perfect, and very few non-English apps might get past our filter, but this seems good enough at this point in our analysis — we shouldn't spend too much time on optimization at this point.

Below, we use the `is_english()` function to filter out the non-English apps for both data sets:

In [20]:
android_cleaned=[]
ios_cleaned=[]
for app in android_clean:
    name=app[0]
    if is_english(name):
        android_cleaned.append(app)
for app in ios:
    name=app[1]
    if is_english(name):
        ios_cleaned.append(app)
        
explore_data(android_cleaned, 0,3,True)
print('\n')
explore_data(ios_cleaned, 0,3,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

We can see that we're left with **9,614** Android apps and **6,183** iOS apps.

# Isolating the Free Apps

As we mentioned in the introduction, we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our data sets contain both free and non-free apps, and we'll need to isolate only the free apps for our analysis. 
Below, we isolate the free apps for both our data sets.

In [21]:
android_final=[]
ios_final=[]
for app in android_cleaned:
    price=app[7]
    if price=='0':
        android_final.append(app)
        
for app in ios_cleaned:
    price=float(app[4])
    if price==0:
        ios_final.append(app)        
        
explore_data(android_final, 0,3,True)
print('\n')
explore_data(ios_final, 0,3,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 8864
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

Finally, we're left with **8,864** Android apps and **3,222** iOS apps, which should be enough for our analysis.

The cleaning the data we made before was comprised of 4 steps, so we:
* Removed inaccurate data
* Removed duplicate app entries
* Removed non-English apps
* Isolated the free apps

# The Analysis

## Most Common Apps by Genre

As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets.

Let's begin the analysis by getting a sense of what are the most common **genres** for each market. 

In [22]:
print(android_header)
print('\n')
explore_data(android_final, 0,2)
print('\n')
print(ios_header)
print('\n')
explore_data(ios_final, 0,2)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']




['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0'

Due to the above, it seems 2 columns as: *'Category'* and *'Genres'* for `android_final` data set and 1 column *'prime_genre'* for `ios_final` data set will be useful for further analysis.

We'll build two functions we can use to analyze the frequency tables:

* One function to generate frequency tables that show percentages
* Another function we can use to display the percentages in a descending order

In [23]:
def freq_table(dataset, index):
    table={}
    total=0
    for app in dataset:
        total+=1
        key=str(app[index])
        if key in table:
            table[key]+=1
        else:
            table[key]=1
            
    for each_key in table:
        table[each_key]=(table[each_key]/total)*100
       
    return table

In [24]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

We will start our analysis by examining the frequency table for the `prime_genre` column of the App Store data set.

In [25]:
display_table(ios_final, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


We can see, that the most popular free English apps on the App Store are *Games* with the more than a half - **58%** of the whole free English App Store data set. *Entertainment* apps are close to **8%**, followed by *Photo and Video* apps, which are close to **5%**.

Overall there **23** different genres among free English ios apps. First 5 genres cover almost **80%** of this defined market with *Education* apps on the fourth place: **3.66%** and *Social Networking* genre on the fifth: **3.3%**.

The general impression is that most of the free English ios apps are designed for fun purposes like games, entertainment, photo and video, social networking, sports, music, while apps with practical goals (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. 

As for now, we can say only about app's genres which are the most numerous, but it does not mean that they have the greatest number of users.

Let's continue by examining the Genres and Category columns of the Google Play data set.

In [26]:
display_table(android_final, 1) # Category

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

Here, 3 the most numerous caterogies among free-english Google Play apps are: *Family* - **19%**, *Game* - **10%** and *Tools* - **8.5%**. Only about **45%** of overall free-english Google Play apps is shared among first 5 categories of overall 33.

Also here practical apps seem to have a better representation on Google Play compared to App Store. This picture is also confirmed by the frequency table we see for the `Genres` column for Android apps below.

In [27]:
display_table(android_final, -4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

There are many more genres for Android Play market then for App Store and the first 5 the most numerous genres cover less than **30%** of overall free-english Android apps market. Where on the first place is *Tools* with **8.5%**.

We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

# Most Popular Apps by Genre on the App Store

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the `Installs` column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the `rating_count_tot` app.

Below, we calculate the average number of user ratings per app genre on the App Store:

In [28]:
unique_genres={}
for app in ios_final:
    key=app[-5]
    if key in unique_genres:
        unique_genres[key]+=1
    else:
        unique_genres[key]=1   
        
avg_num_rating={}
for genre in unique_genres:
    total=0
    len_genre=0
    for app in ios_final:
        genre_app=app[-5] 
        if genre_app==genre:
            total+=float(app[5])
            len_genre+=1
    avg_num=total/len_genre
    avg_num_rating[genre]=avg_num


def display_table2(table):
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
        
display_table2(avg_num_rating)
print('\n')
display_table(ios_final, -5)

Navigation : 86090.33333333333
Reference : 74942.11111111111
Social Networking : 71548.34905660378
Music : 57326.530303030304
Weather : 52279.892857142855
Book : 39758.5
Food & Drink : 33333.92307692308
Finance : 31467.944444444445
Photo & Video : 28441.54375
Travel : 28243.8
Shopping : 26919.690476190477
Health & Fitness : 23298.015384615384
Sports : 23008.898550724636
Games : 22788.6696905016
News : 21248.023255813954
Productivity : 21028.410714285714
Utilities : 18684.456790123455
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Business : 7491.117647058823
Education : 7003.983050847458
Catalogs : 4004.0
Medical : 612.0


Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732


Comparing the frequency of the app's genres with their popularity among the users - the average number of user ratings, we can notice that the most numerous genre as *Games* does not have the biggest number of users.
The most popular apps among the users are in the rare app's genres of the Free English Apple Store market such as *Navigation*: **86K**, 
*Reference*: **75K** and *Social Networking*: **71,5K**.
Let's check apps in these 3 genres:

In [29]:
for app in ios_final:
    if app[-5]=='Navigation':
        print(app[1], ':', app[5] ) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


As we see, spreading between numbers of ratings among the apps is very large, the highest popularity this genre has thanks to the 2 popular apps as Waze and Google Maps.

In [30]:
for app in ios_final:
    if app[-5]=='Reference':
        print(app[1], ':', app[5] )

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


In [31]:
for app in ios_final:
    if app[-5]=='Social Networking':
        print(app[1], ':', app[5] )

Facebook : 2974676
Pinterest : 1061624
Skype for iPhone : 373519
Messenger : 351466
Tumblr : 334293
WhatsApp Messenger : 287589
Kik : 260965
ooVoo – Free Video Call, Text and Voice : 177501
TextNow - Unlimited Text + Calls : 164963
Viber Messenger – Text & Call : 164249
Followers - Social Analytics For Instagram : 112778
MeetMe - Chat and Meet New People : 97072
We Heart It - Fashion, wallpapers, quotes, tattoos : 90414
InsTrack for Instagram - Analytics Plus More : 85535
Tango - Free Video Call, Voice and Chat : 75412
LinkedIn : 71856
Match™ - #1 Dating App. : 60659
Skype for iPad : 60163
POF - Best Dating App for Conversations : 52642
Timehop : 49510
Find My Family, Friends & iPhone - Life360 Locator : 43877
Whisper - Share, Express, Meet : 39819
Hangouts : 36404
LINE PLAY - Your Avatar World : 34677
WeChat : 34584
Badoo - Meet New People, Chat, Socialize. : 34428
Followers + for Instagram - Follower Analytics : 28633
GroupMe : 28260
Marco Polo Video Walkie Talkie : 27662
Miitomo : 2

The same pattern we can notice for the other 2 genres when the average is raised due to several top-famous apps, giants like Facebook, Pinterest, Skype, etc. Our aim is to find popular genres among the apps which are not under famous brands like Pandora, Spotify, and Shazam for music genre. We need a popular genre for new-entred apps. 

If we investigate **Social Networking** genre more attentively, we can notice some unknown apps with a high number of ratings like free video and text messengers or chats with dating features. It seems this niche can be attractive to our goal.

Now let's analyze the Google Play market a bit.

# Most Popular Apps by Genre on Google Play

We have data about the number of installs for the Google Play market, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.)

In [32]:
display_table(android_final, 5) # the Installs columns

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


We will remove '+' sign and will consider the numbers as at least numbers of installs, so an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

Let's calculate the average number of installs per app category for the Google Play data set. 

In [33]:
unique_cats={}
for app in android_final:
    key=app[1]
    if key in unique_cats:
        unique_cats[key]+=1
    else:
        unique_cats[key]=1   
        
avg_instals={}
for category in unique_cats:
    total=0
    len_category=0
    for app in android_final:
        category_app=app[1] 
        if category_app==category:
            instals=app[5]
            instals=instals.replace('+', '')
            instals=instals.replace(',', '')
            total+=float(instals)
            len_category+=1
    avg_num=total/len_category
    avg_instals[category]=avg_num

        
display_table2(avg_instals)
print('\n')
display_table(android_final, 1)

COMMUNICATION : 38456119.167247385
VIDEO_PLAYERS : 24727872.452830188
SOCIAL : 23253652.127118643
PHOTOGRAPHY : 17840110.40229885
PRODUCTIVITY : 16787331.344927534
GAME : 15588015.603248259
TRAVEL_AND_LOCAL : 13984077.710144928
ENTERTAINMENT : 11640705.88235294
TOOLS : 10801391.298666667
NEWS_AND_MAGAZINES : 9549178.467741935
BOOKS_AND_REFERENCE : 8767811.894736841
SHOPPING : 7036877.311557789
PERSONALIZATION : 5201482.6122448975
WEATHER : 5074486.197183099
HEALTH_AND_FITNESS : 4188821.9853479853
MAPS_AND_NAVIGATION : 4056941.7741935486
FAMILY : 3695641.8198090694
SPORTS : 3638640.1428571427
ART_AND_DESIGN : 1986335.0877192982
FOOD_AND_DRINK : 1924897.7363636363
EDUCATION : 1833495.145631068
BUSINESS : 1712290.1474201474
LIFESTYLE : 1437816.2687861272
FINANCE : 1387692.475609756
HOUSE_AND_HOME : 1331540.5616438356
DATING : 854028.8303030303
COMICS : 817657.2727272727
AUTO_AND_VEHICLES : 647317.8170731707
LIBRARIES_AND_DEMO : 638503.734939759
PARENTING : 542603.6206896552
BEAUTY : 51315

The most popular Category-genre among Free-English apps on the Google Play market is **Communication** genre with **38.5** millions numbers of installs.

In [34]:
for app in android_final:
    if app[1]=='COMMUNICATION':
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
Messenger for SMS : 10,000,000+
My Tele2 : 5,000,000+
imo beta free calls and text : 100,000,000+
Contacts : 50,000,000+
Call Free – Free Call : 5,000,000+
Web Browser & Explorer : 5,000,000+
Browser 4G : 10,000,000+
MegaFon Dashboard : 10,000,000+
ZenUI Dialer & Contacts : 10,000,000+
Cricket Visual Voicemail : 10,000,000+
TracFone My Account : 1,000,000+
Xperia Link™ : 10,000,000+
TouchPal Keyboard - Fun Emoji & Android Keyboard : 10,000,000+
Skype Lite - Free Video Call & Chat : 5,000,000+
My magenta : 1,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Seznam.cz : 1,000,000+
Antillean Gold Telegram (original version) : 100,000+
AT&T Visual Voicemail : 10,000,000+
GMX Mail : 10,000,000+
Omlet Chat : 10,000,000+
My Vodacom SA : 5,000,000+
Microsoft Edge : 5,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Calls & Text by Mo+ : 5,000,000+
free 

The same pattern is here too, when average is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts).

Let's explore the sample apps with the non-highest number of installs in this category.

In [35]:
for app in android_final:
    if app[1]=='COMMUNICATION' and (app[5]== '100,000,000+'
                               or app[5]== '50,000,000+'
                               or app[5]== '10,000,000+'):
        print(app[0], ':', app[5])

Messenger for SMS : 10,000,000+
imo beta free calls and text : 100,000,000+
Contacts : 50,000,000+
Browser 4G : 10,000,000+
MegaFon Dashboard : 10,000,000+
ZenUI Dialer & Contacts : 10,000,000+
Cricket Visual Voicemail : 10,000,000+
Xperia Link™ : 10,000,000+
TouchPal Keyboard - Fun Emoji & Android Keyboard : 10,000,000+
Android Messages : 100,000,000+
AT&T Visual Voicemail : 10,000,000+
GMX Mail : 10,000,000+
Omlet Chat : 10,000,000+
free video calls and chat : 50,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
chomp SMS : 10,000,000+
Glide - Video Chat Messenger : 10,000,000+
Text SMS : 10,000,000+
Talkray - Free Calls & Texts : 10,000,000+
GroupMe : 10,000,000+
Firefox Browser fast & private : 100,000,000+
Dolphin Browser - Fast, Private & Adblock🐬 : 50,000,000+
Contacts+ : 10,000,000+
ExDialer - Dialer & Contacts : 10,000,000+
Hiya - Caller ID & Block : 10,000,000+
Mr. Number-Block calls & spam : 10,000,000+
Calls Blacklist - Call Blocker : 10,

Seems **Communication** genre can be the winner for Android market too, above there are a lot of unknown apps with high number of installs, seems the best resolution will be free video and text messanger or chat with dating or gaming feature.

# Conclusions

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that profitable for both the Google Play and the App Store markets may become free video and text messenger or chat with dating or gaming features. People still looking for communication and meeting apps, it can be an international dating app as well as chat for selected popular internet games.